main/FatesHistoryInterfaceMod.F90

```diff
@@ -4707,108 +4931,136 @@ subroutine update_history_hifrq(this,nc,nsites,sites,bc_in,dt_tstep)
-                     ! calculate where we are on multiplexed dimensions
-                     cnlfpft_indx = ileaf + (ican-1) * nlevleaf + (ipft-1) * nlevleaf * nclmax
-                     cnlf_indx = ileaf + (ican-1) * nlevleaf
-                     !
-                     ! first do all the canopy x leaf x pft calculations
-                     hio_parsun_z_si_cnlfpft(io_si,cnlfpft_indx) = hio_parsun_z_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%ed_parsun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_parsha_z_si_cnlfpft(io_si,cnlfpft_indx) = hio_parsha_z_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%ed_parsha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     !
-                     hio_laisun_z_si_cnlfpft(io_si,cnlfpft_indx) = hio_laisun_z_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%ed_laisun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_laisha_z_si_cnlfpft(io_si,cnlfpft_indx) = hio_laisha_z_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%ed_laisha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     !
-                     hio_fabd_sun_si_cnlfpft(io_si,cnlfpft_indx) = hio_fabd_sun_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%fabd_sun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_fabd_sha_si_cnlfpft(io_si,cnlfpft_indx) = hio_fabd_sha_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%fabd_sha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_fabi_sun_si_cnlfpft(io_si,cnlfpft_indx) = hio_fabi_sun_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%fabi_sun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_fabi_sha_si_cnlfpft(io_si,cnlfpft_indx) = hio_fabi_sha_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%fabi_sha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     !
-                     hio_parprof_dir_si_cnlfpft(io_si,cnlfpft_indx) = hio_parprof_dir_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%parprof_pft_dir_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_parprof_dif_si_cnlfpft(io_si,cnlfpft_indx) = hio_parprof_dif_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%parprof_pft_dif_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     !
-                     ! summarize across all PFTs
-                     hio_parsun_z_si_cnlf(io_si,cnlf_indx) = hio_parsun_z_si_cnlf(io_si,cnlf_indx) + &
-                          cpatch%ed_parsun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_parsha_z_si_cnlf(io_si,cnlf_indx) = hio_parsha_z_si_cnlf(io_si,cnlf_indx) + &
-                          cpatch%ed_parsha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
+            
+            if_zenith: if(cpatch%solar_zenith_flag) then
```

Comment:
I'll take a close look through here

Reply:
I see what you mean.

All history variables are initialized to zero as a default, via a call like this:

https://github.com/NGEET/fates/blob/3966cc85f3308a1653db108ef153edca9dcec012/main/FatesHistoryInterfaceMod.F90#L4794

But, If we want to set a value to be ignored from the temporal mean, then we need to explicitly override that zero with an ignore value. This is very often the case for variables with an extra dimension, when certain fields in that dimension just aren't present, like if we don't have any cohorts in a certain size class, we don't want to report zero, we want to report unknown.

Here is an example:

https://github.com/NGEET/fates/blob/3966cc85f3308a1653db108ef153edca9dcec012/main/FatesHistoryInterfaceMod.F90#L5042C68-L5042C86

So I see that in your example, we do have some variables that use an extra dimension, such as canopy layer. Instead of setting them all to zero if cosz<=0, we should make sure they get zero's when appropriate, and the ignore value when appropriate (such as if an index has no representatives).

Reply:
so I reworked the history diagnostics code in this section quite a bit. It no longer relies on the zenith angle for filtering calculations, and does include zeros when there is no atmospheric radiation. It also makes sure that the ignore value is used for indices of multiplexed arrays that are invalid (for example in a leaf layer that is not occupied by any vegetation).